Add Brainpool Curves and Expand ECDSA support in openpgp

0xhaven opened the pull request:
To address https://github.com/keybase/client/issues/1310 (inability to import keys from Brainpool curves)

Commit:
Add Brainpool t1/r1 P-256, P-384 and P-512 curves

Commit:
Use brainpool curves in openpgp

0xhaven commented on brainpool/brainpool.go:
```diff
@@ -0,0 +1,65 @@
+func initP512r1() {
+	p512r1 = &elliptic.CurveParams{Name: "brainpoolP512r1"}
+	p512r1.P, _ = new(big.Int).SetString("AADD9DB8DBE9C48B3FD4E6AE33C9FC07CB308DB3B3C9D20ED6639CCA703308717D4D9B009BC66842AECDA12AE6A380E62881FF2F2D82C68528AA6056583A48F3", 16)
+	p512r1.N, _ = new(big.Int).SetString("AADD9DB8DBE9C48B3FD4E6AE33C9FC07CB308DB3B3C9D20ED6639CCA70330870553E5C414CA92619418661197FAC10471DB1D381085DDADDB58796829CA90069", 16)
```
I'm don't think these `N` values are set correctly.

maxtaco replied:
My reference is here: https://github.com/keybase/kbpgp/blob/master/src/ecc/curves.iced#L149-L179

0xhaven replied:
Yes, this matches that, although the [elliptic.CurveParams](https://golang.org/pkg/crypto/elliptic/#CurveParams) struct has no `A` member (it's assumed to be `-3`)

maxtaco commented:
How about some tests?

See https://github.com/keybase/kbpgp/blob/master/test/files/brainpool256.iced, https://github.com/keybase/kbpgp/blob/master/test/files/brainpool384.iced and https://github.com/keybase/kbpgp/blob/master/test/files/brainpool512.iced

Commit:
Brainpool pgp key parsing tests

0xhaven commented:
@maxtaco I've changed how the r curves are generated (using the t curves as twist), necessary because `"crypto/elliptic"` assumes curves are in short-form Weierstrass with `A = -3`.
I've added basic ECDSA tests and OpenPGP parsing tests for those test keys.

0xhaven commented on brainpool/rcurve.go:
```diff
@@ -0,0 +1,84 @@
+	var curve rcurve
+
+	curve.twisted = twisted
+	curve.params = *twisted.Params()
```
Should probably set `curve.params.Name` here.

maxtaco commented:
@jacobhaven, can you also provide an End-to-End sign/verify and decrypt/encrypt test?

0xhaven commented:
[packet#Signature.Sign](https://github.com/keybase/go-crypto/blob/master/openpgp/packet/signature.go#L536) doesn't appear to support `PubKeyAlgoECDSA`

``` Go
    sig := &packet.Signature{
        PubKeyAlgo: packet.PubKeyAlgoECDSA,
        Hash:       crypto.SHA256,
    }

    h := sig.Hash.New()
    if _, err := h.Write([]byte("Hello World")); err != nil {
        return err
    }

    if err := sig.Sign(h, priv, nil); err != nil {
        return err // "openpgp: unsupported feature: public key algorithm: 19"
    }
```

maxtaco commented:
That is a slight bummer. Maybe just for now verify a sig that was signed
elsewhere?

On Wednesday, March 9, 2016, Jacob H. Haven notifications@github.com
wrote:

> packet#Signature.Sign
> https://github.com/keybase/go-crypto/blob/master/openpgp/packet/signature.go#L536
> doesn't appear to support PubKeyAlgoECDSA
> 
> sig := &packet.Signature{
>         PubKeyAlgo: packet.PubKeyAlgoECDSA,
>         Hash:       crypto.SHA256,
>     }
> 
> ```
> h := sig.Hash.New()
> if _, err := h.Write([]byte("Hello World")); err != nil {
>     return err
> }
> 
> if err := sig.Sign(h, priv, nil); err != nil {
>     return err // "openpgp: unsupported feature: public key algorithm: 19"
> }
> ```
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/keybase/go-crypto/pull/18#issuecomment-194429439.

maxtaco commented:
Or does that implicitly happen in key import?

On Wednesday, March 9, 2016, Maxwell Krohn themax@gmail.com wrote:

> That is a slight bummer. Maybe just for now verify a sig that was signed
> elsewhere?
> 
> On Wednesday, March 9, 2016, Jacob H. Haven <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
> 
> > packet#Signature.Sign
> > https://github.com/keybase/go-crypto/blob/master/openpgp/packet/signature.go#L536
> > doesn't appear to support PubKeyAlgoECDSA
> > 
> > sig := &packet.Signature{
> >         PubKeyAlgo: packet.PubKeyAlgoECDSA,
> >         Hash:       crypto.SHA256,
> >     }
> > 
> > ```
> > h := sig.Hash.New()
> > if _, err := h.Write([]byte("Hello World")); err != nil {
> >     return err
> > }
> > 
> > if err := sig.Sign(h, priv, nil); err != nil {
> >     return err // "openpgp: unsupported feature: public key algorithm: 19"
> > }
> > ```
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/keybase/go-crypto/pull/18#issuecomment-194429439.

0xhaven commented:
I'm trying a different route via https://godoc.org/github.com/keybase/go-crypto/openpgp#DetachSign but that requires parsing a whole `Entity`, not just a private key.

maxtaco commented:
BTW, another option is to add support for signing in `signature.go`.  It might not be too much heavy-lifting.

Commit:
Add full ECDSA PrivateKey support to openpgp

0xhaven commented:
This package was in a weird in-between state wrt ECDSA support. I went through and made it a first-class algorithm everywhere.

maxtaco commented:
This looks outstanding.  Can you add one last test which is to verify a
brainpool sig made with gpg or possibly the JavaScript library?

On Wednesday, March 9, 2016, Jacob H. Haven notifications@github.com
wrote:

> This package was in a weird in-between state wrt ECDSA support. I went
> through and made it a first-class algorithm everywhere.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/keybase/go-crypto/pull/18#issuecomment-194544677.

maxtaco commented on brainpool/rcurve.go:
```diff
@@ -0,0 +1,81 @@
+
+func (curve *rcurve) ScalarBaseMult(scalar []byte) (x, y *big.Int) {
+	return curve.fromTwisted(curve.twisted.ScalarBaseMult(scalar))
+}
```
can you add an assertion of the form:

``` go
var _ elliptic.Curve = (*rcurve)(nil)
```

I find such things useful for readability.

0xhaven replied:
Done.

maxtaco commented:
A small thing and a few more test cases...

Also, @oconnor663 or @akalin-keybase do you want to take a pass through the crypto? Nice to have a 3rd set of eyes. 

Looks great overall!

Commit:
Add test verification of brainpool signatures from GPG

0xhaven commented:
*added test of brainpool signatures generated by GPG*

maxtaco commented:
I'm a :+1:, @oconnor663 said he'll take a pass too, thanks!

oconnor663 commented on brainpool/rcurve.go:
```diff
@@ -0,0 +1,83 @@
+
+func (curve *rcurve) IsOnCurve(x, y *big.Int) bool {
+	return curve.twisted.IsOnCurve(curve.toTwisted(x, y))
+}
```
Is this the part that protects us from ["invalid curve attacks"](https://safecurves.cr.yp.to/twist.html)? I know basically nothing about EC, other than that these scary things exist :p

0xhaven replied:
Yeah, it's checked when [Unmarshaling a point](https://golang.org/src/crypto/elliptic/elliptic.go#L330) onto a curve

oconnor663 commented on brainpool/brainpool_test.go:
```diff
@@ -0,0 +1,49 @@
+
+	if !ecdsa.Verify(&priv.PublicKey, msg, r, s) {
+		t.Fatal("signature didn't verify.")
+	}
```
Do we have any test cases that check that bad signatures fail?

0xhaven replied:
Any suggestions on how to generate a bad signature that could conceivably succeed?

oconnor663 commented:
Just the two comments above that I could think of. Thanks Jacob! At some point I should get you to teach me this stuff :)

maxtaco commented:
I'm going to give a :+1: that overrides @oconnor663's request for test if that test is hard to produce.

akalin-keybase commented on brainpool/rcurve.go:
```diff
@@ -0,0 +1,83 @@
+	three = big.NewInt(3)
+)
+
+func newrcurve(twisted elliptic.Curve, params *elliptic.CurveParams, z *big.Int) *rcurve {
```
noob question: https://golang.org/pkg/crypto/elliptic/#CurveParams says that the `CurveParams` implementation of `Curve` is non-constant time, and we always pass those to `newrcurve`. Is it not a big deal that these curves don't use constant-time ops?

0xhaven replied:
For the NIST curves, Go only has constant time implementations for P-224 a P-256.
There may be potential for timing attacks here, but they would also apply to NIST P-384 and P-521.

maxtaco replied:
Yay NaCl

On Thursday, March 10, 2016, Jacob H. Haven notifications@github.com
wrote:

> In brainpool/rcurve.go
> https://github.com/keybase/go-crypto/pull/18#discussion_r55760611:
> 
> > -   twisted elliptic.Curve
> > -   params  *elliptic.CurveParams
> > -   z       *big.Int
> > -   zinv    *big.Int
> > -   z2      *big.Int
> > -   z3      *big.Int
> > -   zinv2   *big.Int
> > -   zinv3   *big.Int
> >   +}
> >   +
> >   +var (
> > -   two   = big.NewInt(2)
> > -   three = big.NewInt(3)
> >   +)
> >   +
> >   +func newrcurve(twisted elliptic.Curve, params *elliptic.CurveParams, z *big.Int) *rcurve {
> 
> For the NIST curves, Go only has constant time implementations for P-224 a
> P-256.
> There may be potential for timing attacks here, but they would also apply
> to NIST P-384 and P-521.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/keybase/go-crypto/pull/18/files#r55760611.

akalin-keybase replied:
Got it, thanks!

akalin-keybase commented:
:+1:

ebfe commented:
If you copy/paste from github.com/ebfe/brainpool you might copy the tests as well...

maxtaco commented:
@ebfe what is the license on the code in that repo?  @jacobhaven please make an attribution in the comments. And/or is it possible to import @ebfe's library?

ebfe commented:
> @ebfe what is the license on the code in that repo?
> Well there isn't any license. Not putting a license/readme in the repo was an
> (now obviously failed) attempt at discouraging people to use it for anything
> serious. The code was never reviewed by anybody. In particular the
> implementation of the r-curves is quite hackish (crypto/elliptic assumes all
> curves are short-form weierstrass with a=-3 but the r-curves are not), and I'm
> a bit surprised that didn't come up in this review.

If it helps I can put a BSD-style license on the code and would be happy if
eventual fixes/improvements made their way back home.

akalin-keybase commented:
@ebfe the a=-3 fact did come up in this review -- see https://github.com/keybase/go-crypto/pull/18#issuecomment-194295022 .

maxtaco commented on brainpool/brainpool.go:
```diff
@@ -0,0 +1,134 @@
+// Package brainpool implements Brainpool elliptic curves.
+// Implementation of rcurves is borrowed from github.com/ebfe/brainpool
```
remove word "borrowed"

maxtaco commented:
I'm a :+1:. @ebfe thank you for your comments and sharing your code with us.  Sorry for the earlier snub, that wasn't cool on our part.
